In [1]:
import tensorflow as tf
old_v = tf.logging.get_verbosity()
tf.logging.set_verbosity(tf.logging.ERROR)

<h2>Extract MNIST data</h2>
<p style="font-size:20px">You can change the option of one_hot encoding.

In [2]:
from tensorflow.examples.tutorials.mnist import input_data
#get mnist data, with one_hot encoding
mnist = input_data.read_data_sets("MNIST_data/",one_hot=True)
#suppress warnings
tf.logging.set_verbosity(old_v)

Extracting MNIST_data/train-images-idx3-ubyte.gz
Extracting MNIST_data/train-labels-idx1-ubyte.gz
Extracting MNIST_data/t10k-images-idx3-ubyte.gz
Extracting MNIST_data/t10k-labels-idx1-ubyte.gz


In [3]:
num_train = mnist.train.num_examples #55,000
num_validation = mnist.validation.num_examples #5000
num_test = mnist.test.num_examples #10,000

<h2>Test</h2>

In [4]:
#learning rate
lr = 0.01
#number of traning steps
num_steps = 2000
#number of batch_size
batch_size = 128

#network parameters
n_hidden_1 = 500
n_hidden_2 = 300
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

#define a neural net model
def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.sigmoid(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.sigmoid(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_steps):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
    
    print("Training finished!")
    print("learning rate:", lr, "activation functions: Sigmoid", "n_hidden_1:", n_hidden_1,
      "n_hidden_2:", n_hidden_2, "iterations:", num_steps, "batch size:", batch_size)
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

Instructions for updating:
Colocations handled automatically by placer.
Training finished!
learning rate: 0.01 activation functions: Sigmoid n_hidden_1: 500 n_hidden_2: 300 iterations: 2000 batch size: 128
Testing Accuracy: 0.9399


In [5]:
#learning rate
lr = 0.01
#number of traning steps
num_steps = 2000
#number of batch_size
batch_size = 128

#network parameters
n_hidden_1 = 500
n_hidden_2 = 300
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

#define a neural net model
def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.nn.relu(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.nn.relu(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_steps):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
    
    print("Training finished!")
    print("learning rate:", lr, "activation functions: ReLU", "n_hidden_1:", n_hidden_1,
      "n_hidden_2:", n_hidden_2, "iterations:", num_steps, "batch size:", batch_size)
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

Training finished!
learning rate: 0.01 activation functions: ReLU n_hidden_1: 500 n_hidden_2: 300 iterations: 2000 batch size: 128
Testing Accuracy: 0.9591


In [6]:
#learning rate
lr = 0.01
#number of traning steps
num_steps = 20000
#number of batch_size
batch_size = 128

#network parameters
n_hidden_1 = 300
n_hidden_2 = 100
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

#define a neural net model
def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.sigmoid(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.sigmoid(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_steps):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
    
    print("Training finished!")
    print("learning rate:", lr, "activation functions: Sigmoid", "n_hidden_1:", n_hidden_1,
      "n_hidden_2:", n_hidden_2, "iterations:", num_steps, "batch size:", batch_size)
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

Training finished!
learning rate: 0.01 activation functions: Sigmoid n_hidden_1: 300 n_hidden_2: 100 iterations: 20000 batch size: 128
Testing Accuracy: 0.9684


In [7]:
#learning rate
lr = 0.01
#number of traning steps
num_steps = 20000
#number of batch_size
batch_size = 128

#network parameters
n_hidden_1 = 300
n_hidden_2 = 100
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

#define a neural net model
def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.nn.relu(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.nn.relu(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_steps):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
    
    print("Training finished!")
    print("learning rate:", lr, "activation functions: ReLU", "n_hidden_1:", n_hidden_1,
      "n_hidden_2:", n_hidden_2, "iterations:", num_steps, "batch size:", batch_size)
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

Training finished!
learning rate: 0.01 activation functions: ReLU n_hidden_1: 300 n_hidden_2: 100 iterations: 20000 batch size: 128
Testing Accuracy: 0.9533


In [8]:
#learning rate
lr = 0.01
#number of traning steps
num_steps = 20000
#number of batch_size
batch_size = 64

#network parameters
n_hidden_1 = 500
n_hidden_2 = 300
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

#define a neural net model
def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.sigmoid(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.sigmoid(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_steps):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
    
    print("Training finished!")
    print("learning rate:", lr, "activation functions: Sigmoid", "n_hidden_1:", n_hidden_1,
      "n_hidden_2:", n_hidden_2, "iterations:", num_steps, "batch size:", batch_size)
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

Training finished!
learning rate: 0.01 activation functions: Sigmoid n_hidden_1: 500 n_hidden_2: 300 iterations: 20000 batch size: 64
Testing Accuracy: 0.9619


In [9]:
#learning rate
lr = 0.01
#number of traning steps
num_steps = 20000
#number of batch_size
batch_size = 64

#network parameters
n_hidden_1 = 500
n_hidden_2 = 300
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

#define a neural net model
def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.nn.relu(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.nn.relu(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_steps):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
    
    print("Training finished!")
    print("learning rate:", lr, "activation functions: ReLU", "n_hidden_1:", n_hidden_1,
      "n_hidden_2:", n_hidden_2, "iterations:", num_steps, "batch size:", batch_size)
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

Training finished!
learning rate: 0.01 activation functions: ReLU n_hidden_1: 500 n_hidden_2: 300 iterations: 20000 batch size: 64
Testing Accuracy: 0.9434


In [10]:
#learning rate
lr = 0.001
#number of traning steps
num_steps = 2000
#number of batch_size
batch_size = 128

#network parameters
n_hidden_1 = 500
n_hidden_2 = 300
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

#define a neural net model
def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.sigmoid(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.sigmoid(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_steps):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
    
    print("Training finished!")
    print("learning rate:", lr, "activation functions: Sigmoid", "n_hidden_1:", n_hidden_1,
      "n_hidden_2:", n_hidden_2, "iterations:", num_steps, "batch size:", batch_size)
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

Training finished!
learning rate: 0.001 activation functions: Sigmoid n_hidden_1: 500 n_hidden_2: 300 iterations: 2000 batch size: 128
Testing Accuracy: 0.9003


In [11]:
#learning rate
lr = 0.001
#number of traning steps
num_steps = 2000
#number of batch_size
batch_size = 128

#network parameters
n_hidden_1 = 500
n_hidden_2 = 300
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

#define a neural net model
def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.nn.relu(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.nn.relu(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_steps):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
    
    print("Training finished!")
    print("learning rate:", lr, "activation functions: ReLU", "n_hidden_1:", n_hidden_1,
      "n_hidden_2:", n_hidden_2, "iterations:", num_steps, "batch size:", batch_size)
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

Training finished!
learning rate: 0.001 activation functions: ReLU n_hidden_1: 500 n_hidden_2: 300 iterations: 2000 batch size: 128
Testing Accuracy: 0.9298


In [12]:
#learning rate
lr = 0.001
#number of traning steps
num_steps = 20000
#number of batch_size
batch_size = 128

#network parameters
n_hidden_1 = 300
n_hidden_2 = 100
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

#define a neural net model
def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.sigmoid(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.sigmoid(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_steps):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
    
    print("Training finished!")
    print("learning rate:", lr, "activation functions: Sigmoid", "n_hidden_1:", n_hidden_1,
      "n_hidden_2:", n_hidden_2, "iterations:", num_steps, "batch size:", batch_size)
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

Training finished!
learning rate: 0.001 activation functions: Sigmoid n_hidden_1: 300 n_hidden_2: 100 iterations: 20000 batch size: 128
Testing Accuracy: 0.9584


In [13]:
#learning rate
lr = 0.001
#number of traning steps
num_steps = 20000
#number of batch_size
batch_size = 128

#network parameters
n_hidden_1 = 300
n_hidden_2 = 100
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

#define a neural net model
def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.nn.relu(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.nn.relu(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_steps):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
    
    print("Training finished!")
    print("learning rate:", lr, "activation functions: ReLU", "n_hidden_1:", n_hidden_1,
      "n_hidden_2:", n_hidden_2, "iterations:", num_steps, "batch size:", batch_size)
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

Training finished!
learning rate: 0.001 activation functions: ReLU n_hidden_1: 300 n_hidden_2: 100 iterations: 20000 batch size: 128
Testing Accuracy: 0.9568


In [14]:
#learning rate
lr = 0.001
#number of traning steps
num_steps = 20000
#number of batch_size
batch_size = 64

#network parameters
n_hidden_1 = 500
n_hidden_2 = 300
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

#define a neural net model
def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.sigmoid(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.sigmoid(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_steps):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
    
    print("Training finished!")
    print("learning rate:", lr, "activation functions: Sigmoid", "n_hidden_1:", n_hidden_1,
      "n_hidden_2:", n_hidden_2, "iterations:", num_steps, "batch size:", batch_size)
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

Training finished!
learning rate: 0.001 activation functions: Sigmoid n_hidden_1: 500 n_hidden_2: 300 iterations: 20000 batch size: 64
Testing Accuracy: 0.9554


In [15]:
#learning rate
lr = 0.001
#number of traning steps
num_steps = 20000
#number of batch_size
batch_size = 64

#network parameters
n_hidden_1 = 500
n_hidden_2 = 300
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

#define a neural net model
def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.nn.relu(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.nn.relu(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_steps):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
    
    print("Training finished!")
    print("learning rate:", lr, "activation functions: ReLU", "n_hidden_1:", n_hidden_1,
      "n_hidden_2:", n_hidden_2, "iterations:", num_steps, "batch size:", batch_size)
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

Training finished!
learning rate: 0.001 activation functions: ReLU n_hidden_1: 500 n_hidden_2: 300 iterations: 20000 batch size: 64
Testing Accuracy: 0.9614


<h2>Your results</h2>

| learning rate | activation functions | n_hidden_1 | n_hidden_2 | iterations | batch size | accuracy |
|---------------|----------------------|------------|------------|------------|------------|----------|
| 0.01          | Sigmoid              | 500        | 300        | 2000       | 128        | 0.9399   |
| 0.01          | ReLU                 | 500        | 300        | 2000       | 128        | 0.9591   |
| 0.01          | Sigmoid              | 300        | 100        | 20000      | 128        | 0.9684   |
| 0.01          | ReLU                 | 300        | 100        | 20000      | 128        | 0.9533   |
| 0.01          | Sigmoid              | 500        | 300        | 20000      | 64         | 0.9619   |
| 0.01          | ReLU                 | 500        | 300        | 20000      | 64         | 0.9434   |
| 0.001         | Sigmoid              | 500        | 300        | 2000       | 128        | 0.9003   |
| 0.001         | Relu                 | 500        | 300        | 2000       | 128        | 0.9298   |
| 0.001         | Sigmoid              | 300        | 100        | 20000      | 128        | 0.9584   |
| 0.001         | ReLU                 | 300        | 100        | 20000      | 128        | 0.9568   |
| 0.001         | Sigmoid              | 500        | 300        | 20000      | 64         | 0.9554   |
| 0.001         | ReLU                 | 500        | 300        | 20000      | 64         | 0.9614   |

Instead of GradientDescent, I used Adam Optimizer which converges faster. So I choose learning rate 0.01 and 0.001 for my test. Comparing the first six results with the last six resuls, we could see that a larger learning rate has better performance in general. But we also notice that when we use ReLU as the activation function, it performs better with a lower learning rate, as we could see in the last test. I added activation functions and increased the number of iterations which resulted in a great improvement on accuracy, from 0.8 to 0.9 and above. For the batch size and the number of neurons, I don't think it will cause huge difference as we could see in the graph that the accuracy is quite similar. I didn't try to change the number of layers so I can't say the impact. From my persepective, all tests with accuracy above the 96% are trained with 20000 iterations. I think the number of iterations contributes the most to improve the performance, but I also think that this improvement will have a limitation which means that the number of iterations is too large and the outcome remain the same.  